File: 01143. Longest Common Subsequence/solution.go
/**
 * Problem: 1143. Longest Common Subsequence
 *
 * Difficulty: Medium
 *
 * Language: Golang
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

func longestCommonSubsequence(text1 string, text2 string) int {
	// Assign text1 and text2 to smallStr and bigStr respectively
	var smallStr, bigStr string = text1, text2
	// Get the lengths of smallStr and bigStr
	var smallStrLen, bigStrLen int = len(smallStr), len(bigStr)

	// Initialize a DP array to store the length of LCS up to each character in bigStr
	dp := make([]int, bigStrLen)

	// Declare variables for character indices and temporary values
	var charIdxSmall, charIdxBig, prevDP, tempDP int = 0, 0, 0, 0

	// Iterate over each character in smallStr
	for i := 0; i < smallStrLen; i++ {
		// Get the index of the current character in smallStr
		charIdxSmall = int(smallStr[i] - 'a')
		// Store the previous DP value for the current row
		prevDP = 0

		// Iterate over each character in bigStr
		for j := 0; j < bigStrLen; j++ {
			// Get the index of the current character in bigStr
			charIdxBig = int(bigStr[j] - 'a')
			// Temporarily store the current DP value before updating
			tempDP = dp[j]

			// If the characters match, update DP value by adding 1 to prevDP
			if charIdxSmall == charIdxBig {
				dp[j] = prevDP + 1
			// If not matching and not the first character, take the max of current and previous DP values
			} else if j > 0 {
				dp[j] = max(dp[j], dp[j-1])
			}

			// Update prevDP for the next iteration
			prevDP = tempDP
		}
	}
	
	// Return the length of the longest common subsequence
	return dp[bigStrLen-1]
}

File: 01143. Longest Common Subsequence/solution.js
/**
 * Problem: 1143. Longest Common Subsequence
 *
 * Difficulty: Medium
 *
 * Language: JavaScript
 *
 * Performance: Runtime - 10 ms (Beats 100%)
 */

/**
 * Returns the length of the longest common subsequence of two strings
 *
 * @param {string} text1 - First string
 * @param {string} text2 - Second string
 *
 * @returns {number} - Length of the longest common subsequence
 */
const longestCommonSubsequence = (text1, text2) => {
  // Initialize a dynamic programming array to store the length of LCS for each character in text1
  const dp = Array(text1.length).fill(0)
  // Variable to keep track of the longest common subsequence found so far
  let longest = 0

  // Iterate over each character in text2
  for (const charText2 of text2) {
    // Variable to store the current maximum length for the current character in text2
    let currentMaxLength = 0

    // Iterate over each character index in text1
    for (let indexText1 = 0; indexText1 < dp.length; indexText1++) {
      // Update currentMaxLength if a longer subsequence is found
      if (currentMaxLength < dp[indexText1]) currentMaxLength = dp[indexText1]
      // If characters match, update dp for this index and update longest if needed
      else if (charText2 === text1[indexText1]) {
        dp[indexText1] = currentMaxLength + 1
        longest = Math.max(longest, currentMaxLength + 1)
      }
    }
  }

  // Return the length of the longest common subsequence
  return longest
}
